classes/services/UserService.inc.php

```diff
@@ -456,4 +467,137 @@ public function canCurrentUserGossip($userId) {
+		$individualSubscriptionDao = DAORegistry::getDAO('IndividualSubscriptionDAO');
+		$oldUserSubscriptions = $individualSubscriptionDao->getByUserId($userId);
+
+		while ($oldUserSubscription = $oldUserSubscriptions->next()) {
+			$subscriptionJournalId = $oldUserSubscription->getJournalId();
+			$oldUserValidSubscription = $individualSubscriptionDao->isValidIndividualSubscription($userId, $subscriptionJournalId);
+			if ($oldUserValidSubscription) {
+				// Check if new user has a valid subscription for current journal
+				$newUserSubscription = $individualSubscriptionDao->newUserSubscription($mergeIntoUserId, $subscriptionJournalId);
+				if (!$newUserSubscription) {
+					// New user does not have this subscription, transfer old user's
+					$oldUserSubscription->setUserId($mergeIntoUserId);
+					$individualSubscriptionDao->updateObject($oldUserSubscription);
+				} elseif (!$individualSubscriptionDao->isValidIndividualSubscription($mergeIntoUserId, $subscriptionJournalId)) {
+					// New user has a subscription but it's invalid. Delete it and
+					// transfer old user's valid one
+					$individualSubscriptionDao->deleteByUserIdForJournal($mergeIntoUserId, $subscriptionJournalId);
+					$oldUserSubscription->setUserId($mergeIntoUserId);
+					$individualSubscriptionDao->updateObject($oldUserSubscription);
+				}
+			}
+		}
+
+		// Delete any remaining old user's subscriptions not transferred to new user
+		$individualSubscriptionDao->deleteByUserId($userId);
+
+		// Transfer all old user's institutional subscriptions for each journal to
+		// new user. New user now becomes the contact person for these.
+		$institutionalSubscriptionDao = DAORegistry::getDAO('InstitutionalSubscriptionDAO');
+		$oldUserSubscriptions = $institutionalSubscriptionDao->getByUserId($userId);
+
+		while ($oldUserSubscription = $oldUserSubscriptions->next()) {
+			$oldUserSubscription->setUserId($mergeIntoUserId);
+			$institutionalSubscriptionDao->updateObject($oldUserSubscription);
+		}
+
+		// Transfer completed payments.
+		$paymentDao = DAORegistry::getDAO('OJSCompletedPaymentDAO');
+		$paymentFactory = $paymentDao->getByUserId($userId);
+		while ($payment = next($paymentFactory)) {
+			$payment->setUserId($mergeIntoUserId);
+			$paymentDao->updateObject($payment);
+		}
+
+		// Delete the old user and associated info.
+		$sessionDao = DAORegistry::getDAO('SessionDAO');
+		$sessionDao->deleteByUserId($userId);
+		$temporaryFileDao = DAORegistry::getDAO('TemporaryFileDAO');
+		$temporaryFileDao->deleteByUserId($userId);
+		$userSettingsDao = DAORegistry::getDAO('UserSettingsDAO');
+		$userSettingsDao->deleteSettings($userId);
+		$subEditorsDao = DAORegistry::getDAO('SubEditorsDAO');
+		$subEditorsDao->deleteByUserId($userId);
+
+		// Transfer old user's roles
+		$userGroupDao = DAORegistry::getDAO('UserGroupDAO');
+		$userGroups = $userGroupDao->getByUserId($userId);
+		while(!$userGroups->eof()) {
+			$userGroup = $userGroups->next();
```

Comment:
just a mini comment: I think it could be as above while($userGroup = $userGroups->next()), nor?

Reply:
:+1: